File: jupyter_collaboration/ychat.py
# Copyright (c) Jupyter Development Team.
# Distributed under the terms of the Modified BSD License.

# TODO: remove this module in favor of the one in jupyter_ydoc when released.

import json
import jupyter_ydoc
from functools import partial
from typing import Any, Callable, List
from pycrdt import Array, Map


class YChat(jupyter_ydoc.YBaseDoc):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._ydoc["content"] = self._ycontent = Map()
        self._ydoc["messages"] = self._ymessages = Array()

    @property
    def version(self) -> str:
        """
        Returns the version of the document.
        :return: Document's version.
        :rtype: str
        """
        return "1.0.0"

    @property
    def messages(self) -> List:
        return self._ymessages.to_py()

    def get(self) -> str:
        """
        Returns the messages of the document.
        :return: Document's messages.
        :rtype: Any
        """

        messages = self._ymessages.to_py()
        data = dict(messages=messages)
        return json.dumps(data, indent=2, sort_keys=True)

    def set(self, value: str) -> None:
        """
        Sets the content of the document.
        :param value: The content of the document.
        :type value: str
        """
        contents = json.loads(value)
        if "messages" in contents.keys():
            with self._ydoc.transaction():
                for v in contents["messages"]:
                    self._ymessages.append(v)

    def observe(self, callback: Callable[[str, Any], None]) -> None:
        self.unobserve()
        self._subscriptions[self._ystate] = self._ystate.observe(partial(callback, "state"))
        self._subscriptions[self._ymessages] = self._ymessages.observe(
            partial(callback, "messages")
        )
        self._subscriptions[self._ycontent] = self._ycontent.observe(partial(callback, "content"))


# Register the ydoc
jupyter_ydoc.ydocs["jupyter_collaboration.chat_ydoc:YChat"] = YChat

File: packages/docprovider/src/chat/ychat.ts
/*
 * Copyright (c) Jupyter Development Team.
 * Distributed under the terms of the Modified BSD License.
 */

// TODO: remove this module in favor of the one in jupyter_ydoc when released.

import { Delta, DocumentChange, MapChange, YDocument } from '@jupyter/ydoc';
import { JSONExt, JSONObject } from '@lumino/coreutils';
import * as Y from 'yjs';

/**
 * Definition of the shared Chat changes.
 */
export type ChatChange = DocumentChange & {
  /**
   * Messages list change
   */
  messagesChange?: Delta<any>;
  /**
   * Content change
   */
  contentChange?: MapChange;
};

interface IDict<T = any> {
  [key: string]: T;
}

/**
 * The collaborative chat shared document.
 */
export class YChat extends YDocument<ChatChange> {
  /**
   * Create a new collaborative chat model.
   */
  constructor(options?: YDocument.IOptions) {
    super(options);
    this._content = this.ydoc.getMap<IDict>('content');
    this._content.observe(this._contentObserver);

    this._messages = this.ydoc.getArray<IDict>('messages');
    this._messages.observe(this._messagesObserver);
  }

  /**
   * Document version
   */
  readonly version: string = '1.0.0';

  /**
   * Static method to create instances on the sharedModel
   *
   * @returns The sharedModel instance
   */
  static create(options?: YDocument.IOptions): YChat {
    return new YChat(options);
  }

  get content(): JSONObject {
    return JSONExt.deepCopy(this._content.toJSON());
  }

  get messages(): string[] {
    return JSONExt.deepCopy(this._messages.toJSON());
  }

  setMessage(value: IDict): void {
    this._messages.push([value]);
  }

  private _contentObserver = (event: Y.YMapEvent<IDict>): void => {
    this._changed.emit(this.content);
  };

  private _messagesObserver = (event: Y.YArrayEvent<IDict>): void => {
    const changes: ChatChange = {};
    changes.messagesChange = event.delta;
    this._changed.emit(changes);
  };

  private _content: Y.Map<IDict>;
  private _messages: Y.Array<IDict>;
}
